File: cpp/08_039.cpp
#include <vector>
#include <stack>
#include <algorithm>

int maxProductivityZone(std::vector<int>& capacities) {
    std::stack<int> stack;
    int maxArea = 0;
    int n = capacities.size();
    
    for (int i = 0; i < n; i++) {
        // Process stations with higher capacity than current
        while (!stack.empty() && 
               capacities[stack.top()] > capacities[i]) {
            int height = capacities[stack.top()];
            stack.pop();
            int width = stack.empty() ? i : 
                        i - stack.top() - 1;
            maxArea = std::max(maxArea, height * width);
        }
        
        stack.push(i);
    }
    
    // Process remaining stations in stack
    while (!stack.empty()) {
        int height = capacities[stack.top()];
        stack.pop();
        int width = stack.empty() ? n : 
                    n - stack.top() - 1;
        maxArea = std::max(maxArea, height * width);
    }
    
    return maxArea;
}

File: data structures and algorithms/cs/08_039.cs
using System.Collections.Generic;

public class ProductivityZoneCalculator
{
   public static int MaxProductivityZone(int[] capacities)
   {
      var stack = new Stack<int>();
      int maxArea = 0;
      int n = capacities.Length;

      for (int i = 0; i < n; i++)
      {
         while (stack.Count > 0 &&
            capacities[stack.Peek()] > capacities[i])
         {
            int height = capacities[stack.Pop()];
            int width = stack.Count == 0 ? i :
               i - stack.Peek() - 1;
            maxArea = Math.Max(maxArea, height * width);
         }

         stack.Push(i);
      }

      while (stack.Count > 0)
      {
         int height = capacities[stack.Pop()];
         int width = stack.Count == 0 ? n :
            n - stack.Peek() - 1;
         maxArea = Math.Max(maxArea, height * width);
      }

      return maxArea;
   }
}